File: BEMLIB_22.10/01_num_meth/02_lin_calc/lu_c.m
%---
% Crout LU decomposition
%---
n=4;
a=[ 3 2 3 7 ; 3 4 9 5 ; 4 5 6 5 ; -1 2 3 4 ];
%---

  for i=1:n
    u(i,i)=1.0;
    l(i,1)=a(i,1);
    u(1,i)=a(1,i)/l(1,1);
  end

%---
   for k=2:n
%---

   for i=k:n
     l(i,k)=a(i,k);
     for m=1:k-1
       l(i,k) = l(i,k) - l(i,m)*u(m,k);
     end
   end

   for j=k+1:n
      u(k,j) = a(k,j);
      for m=1:k-1
         u(k,j) = u(k,j) - l(k,m)*u(m,j);
      end
      u(k,j) = u(k,j)/l(k,k);
   end

%---
end
%---

a
l
u
l*u

d=diag(diag(l));
lD = l*inv(d)
uD = d*u

File: BEMLIB_22.10/01_num_meth/02_lin_calc/lu_c.f
      subroutine crout (n,a,l,u)

c-----------------------------------------
c Copyright by C. Pozrikidis, 1999
c All rights reserved.
c
c This program is to be used only under the
c stipulations of the licensing agreement
c----------------------------------------

c------------------------------------------------
c This program accompanies the book:
c
c          C. Pozrikidis
c "Numerical Computation in Science and Engineering"
c     Oxford University Press, 1998
c------------------------------------------------

c---------------------
c  Crout decomposition
c
c  Algorithm 2.6.2
c---------------------

      Implicit Double Precision (a-h,o-z)

      Double Precision l(100,100)
      Dimension        u(100,100),a(100,100)

c---------------------------
c first row and first column
c---------------------------

      Do i=1,n
        u(i,i) = 1.0D0
        l(i,1) = a(i,1)
        u(1,i) = a(1,i)/l(1,1)
      End Do

c-------
c others
c-------

      Do k=2,n

        Do i=k,n
         sum=0.0D0
         Do m=1,k-1
           sum = sum + l(i,m)*u(m,k)
         End Do
         l(i,k) = a(i,k) - sum
        End Do

        Do j=k+1,n
          sum = 0.0D0
          Do m=1,k-1
            sum = sum + l(k,m)*u(m,j)
          End Do
          u(k,j) = (a(k,j)-sum)/l(k,k)
        End Do

      End Do

c-----
c Done
c-----

      Return
      End
